platform/linux/mbedtls/network_mbedtls_wrapper.c

```diff
@@ -286,73 +288,120 @@ IoT_Error_t iot_tls_connect(Network *pNetwork, TLSConnectParams *params) {
 	return (IoT_Error_t) ret;
 }
 
+/* When this much time has elapsed after receiving MBEDTLS_ERR_SSL_WANT_READ
+ * or MBEDTLS_ERR_SSL_WANT_WRITE, then iot_tls_write will return
+ * NETWORK_SSL_WRITE_TIMEOUT_ERROR. */
+#ifndef IOT_SSL_WRITE_RETRY_TIMEOUT_MS
+	#define IOT_SSL_WRITE_RETRY_TIMEOUT_MS 10
+#endif
+
 IoT_Error_t iot_tls_write(Network *pNetwork, unsigned char *pMsg, size_t len, Timer *timer, size_t *written_len) {
-	size_t written_so_far;
-	bool isErrorFlag = false;
-	int frags;
+	mbedtls_ssl_context *pSsl = &(pNetwork->tlsDataParams.ssl);
+	size_t txLen = 0U;
 	int ret = 0;
-	TLSDataParams *tlsDataParams = &(pNetwork->tlsDataParams);
+	/* This timer checks for a timeout whenever MBEDTLS_ERR_SSL_WANT_READ
+	 * or MBEDTLS_ERR_SSL_WANT_WRITE are returned by mbedtls_ssl_write.
+	 * Timeout is specified by IOT_SSL_WRITE_RETRY_TIMEOUT_MS. */
+	Timer writeTimer;
 
-	for(written_so_far = 0, frags = 0;
-		written_so_far < len && !has_timer_expired(timer); written_so_far += ret, frags++) {
-		while(!has_timer_expired(timer) &&
-			  (ret = mbedtls_ssl_write(&(tlsDataParams->ssl), pMsg + written_so_far, len - written_so_far)) <= 0) {
-			if(ret != MBEDTLS_ERR_SSL_WANT_READ && ret != MBEDTLS_ERR_SSL_WANT_WRITE) {
-				IOT_ERROR(" failed\n  ! mbedtls_ssl_write returned -0x%x\n\n", -ret);
-				/* All other negative return values indicate connection needs to be reset.
-		 		* Will be caught in ping request so ignored here */
-				isErrorFlag = true;
-				break;
+	/* This variable is unused */
+	(void) timer;
+
+	/* The timer must be started in case no bytes are written on the first try */
+	init_timer(&writeTimer);
+	countdown_ms(&writeTimer, IOT_SSL_WRITE_RETRY_TIMEOUT_MS);
+
+	while(len > 0U) {
+		ret = mbedtls_ssl_write(pSsl, pMsg, len);
+
+		if(ret > 0) {
+			if((size_t) ret > len) {
+				IOT_ERROR("More bytes written than requested\n\n");
+				return NETWORK_SSL_WRITE_ERROR;
 			}
-		}
-		if(isErrorFlag) {
-			break;
-		}
-	}
 
-	*written_len = written_so_far;
+			/* Successfully sent data, so reset the timeout */
+			init_timer(&writeTimer);
+			countdown_ms(&writeTimer, IOT_SSL_WRITE_RETRY_TIMEOUT_MS);
 
-	if(isErrorFlag) {
-		return NETWORK_SSL_WRITE_ERROR;
-	} else if(has_timer_expired(timer) && written_so_far != len) {
-		return NETWORK_SSL_WRITE_TIMEOUT_ERROR;
+			txLen += ret;
+			pMsg += ret;
+			len -= ret;
+		} else if(ret == MBEDTLS_ERR_SSL_WANT_READ ||
+				ret == MBEDTLS_ERR_SSL_WANT_WRITE) {
+			if(has_timer_expired(&writeTimer)) {
+				return NETWORK_SSL_WRITE_TIMEOUT_ERROR;
```

Comment:
How about passing the length that is read here back to the calling code so the calling code has a flexibility to decide on whether it needs a retry or a failure?